lib/cli-engine.js

```diff
@@ -372,30 +400,60 @@ CLIEngine.prototype = {
     executeOnFiles: function(files) {
 
         var results = [],
-            processed = [],
+            processed = {},
             options = this.options,
             configHelper = new Config(options),
-            ignoredPaths = IgnoredPaths.load(options),
-            exclude = ignoredPaths.contains.bind(ignoredPaths),
+            ignoredPaths = options.ignore ? IgnoredPaths.load(options).patterns : [],
+            extensions = options.extensions.map(function(ext) {
+                return ext.charAt(0) === "." ? ext : "." + ext;
+            }),
+            globOptions,
             stats;
 
-        traverse({
-            files: files,
-            extensions: options.extensions,
-            exclude: options.ignore ? exclude : false
-        }, function(filename) {
+        files = files.map(processPath);
+        ignoredPaths = ignoredPaths.map(processPath);
+
+        globOptions = {
+            nodir: true,
+            realpath: true,
+            ignore: ignoredPaths
+        };
+
+        /**
+         * Executes the linter on a file defined by the `filename`. Skips
+         * unsupported file extensions and any files that are already linted.
+         * @param {string} filename The file to be linted
+         * @returns {void}
+         */
+        function executeOnFile(filename) {
+            if (extensions.indexOf(path.extname(filename)) === -1 || processed[filename]) {
+                return;
+            }
 
             debug("Processing " + filename);
 
-            processed.push(filename);
+            processed[filename] = true;
             results.push(processFile(filename, configHelper));
+        }
+
+        files.forEach(function(pattern) {
+            glob.sync(pattern, globOptions).forEach(executeOnFile);
         });
 
         // only warn for files explicitly passed on the command line
         if (options.ignore) {
             files.forEach(function(file) {
-                if (fs.statSync(path.resolve(file)).isFile() && processed.indexOf(file) === -1) {
-                    results.push(createIgnoreResult(file));
+                // Since `file` can be a glob pattern, we need the try-catch around
+                // `fs.statSync` calls.
+                try {
```

Comment:
It is a string though! :) It just can be a glob string which is not necessarily an actual file or directory path. I've researched other ways but `fs.exists` and `fs.existsSync` are deprecated and the alternatives implemented are exactly this: https://github.com/sindresorhus/path-exists/blob/master/index.js